Tree: Implement Forest proxy

sharptrip opened the pull request:
This is the first iteration of a proxy implementation to showcase simple read/write from/to forest which might be used as a very basic API. Currently only reading from the forest is supported.

### Planned for future

- initial load of data in other formats (e.g. "json >>> JsonableTree<Schema>" )
- write data (requires schema-based data conversion)

sharptrip committed:
Implement Forest proxy

CraigMacomber commented on packages/dds/tree/src/test/forestProxy.spec.ts:
```diff
@@ -0,0 +1,73 @@
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+import { strict as assert } from "assert";
```
I recently added a feature-libraries folder inside `test` since we were getting a lot of top level test files. This should get moved in there.

CraigMacomber commented on packages/dds/tree/src/test/forestProxy.spec.ts:
```diff
@@ -0,0 +1,73 @@
+		},
+	};
+
+	beforeEach(() => {
```
I don't really like the mutable values outside of the test, setup in beforeEach pattern.

I think a more functional style, where we have a "buildTestProxy" function each test calls would be better.

This way it would be easier to add more tests that do different things (ex: allowing the test to select the content for the proxy could be done with an optional parameter) and the data flow is easier to follow.

sharptrip replied:
Done with the function. I intentionally left a test suite structure as it is (i.e. without `testCases` and a corresponding massive processing of those) since currently it tests more the way how the proxy is used rather than the way how the data types are handled as basically we have only two cases: a primitive and a node. It can be adjusted very easily if more complex cases/types will follow.

CraigMacomber commented on packages/dds/tree/src/feature-libraries/forest-proxy/forestProxy.ts:
```diff
@@ -0,0 +1,103 @@
+class TargetForest {
+	private readonly _cursor: ITreeSubscriptionCursor;
+
+	constructor(
```
this unconditionally allocates a cursor which needs to be freed. This should probably be documented (along with the fact that it doesn't modify the provided cursor.

sharptrip replied:
fixed

CraigMacomber commented on packages/dds/tree/src/feature-libraries/forest-proxy/forestProxy.ts:
```diff
@@ -0,0 +1,103 @@
+
+	constructor(
+		public readonly forest: IEditableForest,
+		_cursor?: ITreeSubscriptionCursor,
```
not sure why this has an underscore prefix

DLehenbauer replied:
It's the Fluid convention for handling conflicts between a field that is re-exposed as a property (below).

(Exposing as a readonly field would avoid the conflict.)

CraigMacomber commented on packages/dds/tree/src/feature-libraries/forest-proxy/forestProxy.ts:
```diff
@@ -0,0 +1,103 @@
+export const proxySymbol = Symbol("forest-proxy");
+
+class TargetForest {
+	private readonly _cursor: ITreeSubscriptionCursor;
```
since this is readonly and there is a public getterwith a matching type, why not just make it a public field?

sharptrip replied:
That's a leftover from my experiments.

CraigMacomber commented on packages/dds/tree/src/feature-libraries/forest-proxy/forestProxy.ts:
```diff
@@ -0,0 +1,103 @@
+		if (_cursor) {
+			this._cursor = _cursor.fork();
+		} else {
+			this._cursor = forest.allocateCursor();
+			forest.tryMoveCursorTo(forest.root(forest.rootField), this._cursor);
```
this being the only use of forest in here makes me think maybe I should change the ITreeSubscriptionCursor API a bit so you don't need top reference the forest at all. I should follow up on this.

DLehenbauer replied:
Another option would be to make it mandatory to pass a cursor that is positioned at the root of the subtree you want to proxify.

However, we'll may still need a reference to forest in order to apply edits (unless we want to pass a callback for apply, like `apply: (Delta.Root) => void`).

sharptrip replied:
> this being the only use of forest in here makes me think maybe I should change the ITreeSubscriptionCursor API a bit so you don't need top reference the forest at all. I should follow up on this.

I would keep a forest as an entity, which is clear for the "user", whereas cursors for me are kind of implementation details (at least, in this particular case). On the other hand, it would be nice to have a public reference to the forest the cursor attached to, so that this code could be simplified plus the comment from @DLehenbauer regarding edits.

> Another option would be to make it mandatory to pass a cursor that is positioned at the root of the subtree you want to proxify.

This usecase seems for me to happen quite often ("allocate a cursor and move it to the root"), especially at the very beginning of each and every development in case of new "customers", so why not to add it to the forest API?

DLehenbauer commented on packages/dds/tree/src/feature-libraries/forest-proxy/forestProxy.ts:
```diff
@@ -0,0 +1,103 @@
+			return {
+				configurable: true,
+				enumerable: true,
+				value: target.cursor.value,
```
Did you mean to read the value of the child before returning the cursor back to the parent?

DLehenbauer commented on packages/dds/tree/src/feature-libraries/forest-proxy/forestProxy.ts:
```diff
@@ -0,0 +1,103 @@
+	get: (target: TargetForest, key: string): any => {
+		const result = target.cursor.down(key as FieldKey, 0);
+		if (result === TreeNavigationResult.NotFound) {
+			return Reflect.get(target, key);
```
When the current node lacks the requested key, why is the fallback behavior to reflect on the TargetForest?

Is this perhaps working by coincidence since the Reflect.get(..) will typically return 'undefined' (unless you ask for 'cursor', 'type', etc.)

sharptrip replied:
> When the current node lacks the requested key, why is the fallback behavior to reflect on the TargetForest?
> 
> Is this perhaps working by coincidence since the Reflect.get(..) will typically return 'undefined' (unless you ask for 'cursor', 'type', etc.)

Not really in this place. Currently it is not needed, you are right, but there are ideas to maybe enhance the target with some utilities so that this pattern will provide an access to them.

DLehenbauer commented on packages/dds/tree/src/feature-libraries/forest-proxy/forestProxy.ts:
```diff
@@ -0,0 +1,103 @@
+		configurable: true,
+		writable: false,
+		value: proxySymbol,
+	});
```
FYI - I think this is creating a property on the Proxy that points to itself.  I'm guessing that this is WIP to cache/reuse proxy instances?

sharptrip replied:
> FYI - I think this is creating a property on the Proxy that points to itself. I'm guessing that this is WIP to cache/reuse proxy instances?

Exactly. More or less the same reason as for your comment above regarding reflection.

DLehenbauer commented on packages/dds/tree/src/test/forestProxy.spec.ts:
```diff
@@ -0,0 +1,73 @@
+		assert.deepEqual(proxy.friends, { Mat: "Mat" });
+		assert.deepEqual(Object.keys(proxy.address), ["street", "zip"]);
+		assert.equal(proxy.address.street, "treeStreet");
+	});
```
Awesome to see this working!  :-)

sharptrip committed:
Merge branch 'main' into forest-proxy

sharptrip committed:
follow up comments part 1

sharptrip committed:
fix the fix (underscore convention)

sharptrip committed:
Address comments, add cache, add basic array support

CraigMacomber commented on packages/dds/tree/src/feature-libraries/forest-proxy/forestProxy.ts:
```diff
@@ -0,0 +1,128 @@
+
+class TargetForest {
+	public readonly cursor: ITreeSubscriptionCursor;
+	private readonly _children: Map<string, any>;
```
we should avoid the `any` type and instead use more specific types (especially outside of the internals of functions)

If there really is a reason to use the `any`, the field really needs to document what is allowed to go there.

CraigMacomber commented on packages/dds/tree/src/feature-libraries/forest-proxy/forestProxy.ts:
```diff
@@ -0,0 +1,128 @@
+
+export const proxySymbol = Symbol("forest-proxy");
+
+class TargetForest {
```
This class should have a doc comment. its not clear what it is. Based on only the name sounds like a special kind of forest, but I don't think thats right.

CraigMacomber commented on packages/dds/tree/src/feature-libraries/forest-proxy/forestProxy.ts:
```diff
@@ -0,0 +1,128 @@
+		return this.cursor.type;
+	}
+
+	public getChildProxy(key: string): any {
```
public APIs should not use `any`.

CraigMacomber commented on packages/dds/tree/src/feature-libraries/forest-proxy/forestProxy.ts:
```diff
@@ -0,0 +1,128 @@
+ * It is the only package level export for forestProxy.
+ * @returns a proxy wrapping the given {@link IEditableForest}.
+ */
+export const proxifyForest = (forest: IEditableForest) => proxify(forest);
```
the return type of this function is implicitly a type which is not exported  (TargetForest), not documented, and hard to figure out from looking at the code (you have to read the implementation, then follow that to proxify and read its implementation). We mostly avoid implicit return types in our code.

This also breaks the build (API extractor requires package exported APIs to only use package exported types).

This code should explicitly state a return type, which should be an exported and documented type.

CraigMacomber commented on packages/dds/tree/src/feature-libraries/forest-proxy/README.md:
```diff
@@ -0,0 +1,7 @@
+# Forest Proxy
+
+Implementation of Forest proxy.
```
This sentence doesn't help much. We want to answer, "what is Forest Proxy?"

From the name, I's expect it's an implementation of the forest interfaces done by wrapping a different forest using proxies. I don't think that's what it is.

At the high level, I think what this module is trying to do implement an API for reading and editing trees which are stored in an IEditiableForest that is friendly for schema aware code (aka statically typed with respect to the schema) to use. I have been calling this abstraction `EditableTree`.

I can see two ways we can organize the code for this:
1. A single module/folder that defines the both the abstraction and a single implementation. In this case I think calling it editable-tree makes sense. Proxies are just an implementation detail of that implementation.
2. Two modules. One which defines the abstraction (editable-tree) and a separate one which provides a proxy+forest based implementation of editable-tree, which I would call editable-tree-forest-proxy.

I think the use of proxies is really an implementation detail that shouldn't be in the name.

CraigMacomber commented:
The implementation and tests look great, but I have lots of opinions on this regarding naming, documentation and typing that will likely take a few iterations of review to sort out. I'm happy to do that review, but we could also take a different approach. If you want I could fork this branch, and make a PR back into it with all my opinionated renamed, docs and typing. Let me know what workflow you would prefer.

sharptrip committed:
Merge branch 'main' into forest-proxy

sharptrip committed:
rename artefacts, fix fence, better types

sharptrip committed:
rough array support

sharptrip committed:
finalize array support (technically)

sharptrip committed:
cleanup

sharptrip committed:
update documentation

sharptrip committed:
eslint: remove a trailing space

CraigMacomber commented on packages/dds/tree/src/feature-libraries/editable-tree/README.md:
```diff
@@ -0,0 +1,9 @@
+# Editable Tree
+
+Editable Tree is a simple API prividing an access to a Forest data.
```
```suggestion
Editable Tree is a simple API for accessing a Forest's data.
```

CraigMacomber commented on packages/dds/tree/src/feature-libraries/editable-tree/README.md:
```diff
@@ -0,0 +1,9 @@
+
+Editable Tree is a simple API prividing an access to a Forest data.
+
+An entry point is a `getEditableTree` function which returns a JS Proxy typed as `IEditableTree` allowing to read and write nodes of the Fores in a "JavaScript-object"-like manner, thus hiding all the Forest complexity.
+
+This is intended to showcase basic read and write opeartions from and to the Forest based on a reference `ObjectForest` and a corresponding `Cursor` implementation.
+
+Write operations require a schema-based data format convertion to be adopted/implemented, which will follow.
```
I think the use of proxies is an implementation detail, and it would be nice to explicitly separate the goals of this library from its internals and implementation status. Also nothing in here should be ObjectForest specific.

```suggestion
The entry point is `getEditableTree` which returns an `IEditableTree` which allows reading and writing nodes of the Forest in a "JavaScript-object"-like manner.

## Status

Currently implementation uses a JavaScript proxy and is focused on reading the tree: support for editing is not yet implemented.

Write operations will require a schema-based data format conversion to be implemented.

```

CraigMacomber commented on packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts:
```diff
@@ -0,0 +1,181 @@
+export const proxySymbol = Symbol.for("editable-tree");
+
+/**
+ * A dummy placeholder for a Proxy target.
```
I think this comment is wrong. The target for your proxy is ObjectEditableTree (the target is what is wrapped by the proxy, not the interface the proxy implements).
```suggestion
 * A tree with can be traversed and edited.
 * TODO: support editing.
```

CraigMacomber commented on packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts:
```diff
@@ -0,0 +1,181 @@
+import { IEditableForest, TreeNavigationResult, ITreeSubscriptionCursor } from "../../forest";
+import { TreeSchemaIdentifier } from "../../schema-stored";
+
+export const proxySymbol = Symbol.for("editable-tree");
```
I think searching the runtime for other symbols with this string is not useful here. Just use the constructor and then we can also use unique symbol.

We could also use symbols for accessing other data off nodes like the schema/type of the node:

```suggestion
export const proxySymbol: unique symbol = Symbol("editable-tree proxySymbol");
export const type: unique symbol =Symbol("editable-tree type");
```

CraigMacomber commented on packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts:
```diff
@@ -0,0 +1,181 @@
+/**
+ * A dummy placeholder for a Proxy target.
+ */
+export interface IEditableTree {}
```
This should provide a more specific type. Something more like:

```suggestion
export interface IEditableTree {
    [type]: TreeSchema;
    [proxySymbol]: ObjectEditableTree; // Implementation looks like it claims to have this
    [Symbol.iterator]: Iterator<IEditableTree>; // Not sure this is the right type
    [key: string]: undefined | IEditableTree;
}
```

sharptrip replied:
Still not sure on how the structure, tbd

CraigMacomber commented on packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts:
```diff
@@ -0,0 +1,181 @@
+	return { result, isArray };
+};
+
+function accessArrayCursor(cursor: ITreeSubscriptionCursor) {
```
An explicit return type would make this easier to read.

CraigMacomber commented on packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts:
```diff
@@ -0,0 +1,181 @@
+			const value = cursor.value;
+			cursor.up();
+			yield value;
+		}
```
This is much less efficient that using cursor.seek(1). See mapCursorField for an example.

Also consider not doing this with lazy iteration. If anyone modifies the cursor while this is iterating it will do really strange things. If you keep this pattern, you may want to document that limitation.

Might be better to just replace the whole thing with `return mapCursorField(cursor, (c) => c.value)`.

If keeping this code using up and down, you need to check the returned results. If this encountered "partial" data (thats not downloaded) it would creep up the tree one level after duplicating the value.
mapCursorField handles this correctly so you might want to just use it. I'm also working on revising the cursor API to avoid this pitfall, see https://github.com/microsoft/FluidFramework/pull/11610

sharptrip replied:
Ok, thanks for the info! Will update.

CraigMacomber commented on packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts:
```diff
@@ -0,0 +1,181 @@
+ * A Proxy handler provides a basic read/write access to the Forest by means of the cursors.
+ */
+const handler: ProxyHandler<ObjectEditableTree> = {
+	get: (target: ObjectEditableTree, key: string | symbol): any => {
```
avoid the any type where possible. I think `unknown` should work here:
```suggestion
	get: (target: ObjectEditableTree, key: string | symbol): unknown => {
```

CraigMacomber commented on packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts:
```diff
@@ -0,0 +1,181 @@
+const handler: ProxyHandler<ObjectEditableTree> = {
+	get: (target: ObjectEditableTree, key: string | symbol): any => {
+		if (typeof key === "string") {
+			const { result, isArray } = tryMoveDown(target.cursor, key as FieldKey);
```
You should be able to use `brand` instead of casts for these. It's a little safer (it will error if you try to brand something thats not a string to a FieldKey where as will only error if it does not overlap with the string type)
```suggestion
			const { result, isArray } = tryMoveDown(target.cursor, brand(key));
```

CraigMacomber commented on packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts:
```diff
@@ -0,0 +1,181 @@
+		if (typeof key === "string") {
+			const { result, isArray } = tryMoveDown(target.cursor, key as FieldKey);
+			if (result === TreeNavigationResult.Ok) {
+				const value = target.cursor.value === undefined
+					? target.getChildNode(isArray ? EmptyKey : key as FieldKey)
+					: target.cursor.value;
```
assuming the value isn't ever null (maybe you need to support that here?) you can just do:
```suggestion
				const value = target.cursor.value ?? target.getChildNode(isArray ? EmptyKey : key as FieldKey);
```

sharptrip replied:
I am sure `null` will be needed here. Let's keep it as is.

CraigMacomber commented on packages/dds/tree/src/feature-libraries/editable-tree/fence.json:
```diff
@@ -0,0 +1,5 @@
+{
+    "tags": [ "forest-proxy" ],
+    "exports": ["index"],
+    "imports": ["schema-stored","tree","dependency-tracking","forest","util","rebase","feature-libraries"]
```
I don't think this should depend on rebase or feature-libraries.

```suggestion
    "imports": ["schema-stored","tree","dependency-tracking","forest","util"]
```

sharptrip committed:
Update packages/dds/tree/src/feature-libraries/editable-tree/README.md

Co-authored-by: Craig Macomber (Microsoft) <42876482+CraigMacomber@users.noreply.github.com>

sharptrip committed:
Update packages/dds/tree/src/feature-libraries/editable-tree/README.md

Co-authored-by: Craig Macomber (Microsoft) <42876482+CraigMacomber@users.noreply.github.com>

sharptrip committed:
Update packages/dds/tree/src/feature-libraries/editable-tree/fence.json

Co-authored-by: Craig Macomber (Microsoft) <42876482+CraigMacomber@users.noreply.github.com>

sharptrip committed:
Update packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts

Co-authored-by: Craig Macomber (Microsoft) <42876482+CraigMacomber@users.noreply.github.com>

sharptrip committed:
follow-ups: proxy symbol, brand keys, avoid `any`

sharptrip committed:
"symbolize" proxy target, finalize array support

sharptrip committed:
fix array element descrptor

sharptrip committed:
simplify, cleanup, prepare to rework

sharptrip committed:
Merge branch 'main' into forest-proxy

sharptrip committed:
attempt to implement a proxy type

sharptrip committed:
cleanup, update docs & tests

sharptrip committed:
Merge branch 'main' into forest-proxy

CraigMacomber commented on packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts:
```diff
@@ -0,0 +1,234 @@
+import { brand } from "../../util";
+import { TreeSchemaIdentifier } from "../../schema-stored";
+
+export const proxySymbol: unique symbol = Symbol("editable-tree proxy");
+
+/**
+ * A unique Symbol to get a method of {@link EditableTree} which returns a type of a node or its children.
+ *
+ * One can call `foo[getTypeSymbol]("bar")` to get a type of `bar` or
+ * without a field name i.e. `foo.bar[getTypeSymbol]()` for the same if `bar` is a non-primitive.
+ */
+export const getTypeSymbol: unique symbol = Symbol("editable-tree getType");
+
+/**
+ * This is a basis type for {@link EditableTree}.
+ *
+ * New features like e.g. to get an iterator over node's fields should be defined here
+ * and for each and every new method its own symbol should be created correspondingly.
+ * This prevents accidental mixin with user-defined string keys.
+ */
+export interface IEditableTreeSignature<T> {
+	[key: string]: T extends number | string | boolean ? T : EditableTreeNode<T>;
+	[getTypeSymbol]: (key?: FieldKey) => TreeSchemaIdentifier;
+}
+
+/**
+ * This converts a type T into an {@link EditableTree} node or a primitive.
+ */
+export type EditableTreeNode<T> = {
+	[P in keyof T]?: T[P] extends number | string | boolean ? T[P] : EditableTree<T[P]>;
+};
+
+/**
+ * A tree which can be traversed and edited.
+ * TODO: support editing.
+ */
+export type EditableTree<T> = IEditableTreeSignature<T> & EditableTreeNode<T>;
```
I'm thinking an API like this would work well:

The two symbols might be useful elsewhere as well (If something e else needs them we could move them to the tree folder, though we can leave them here for now).

I've made it non generic: trying to have it do a more specific typed API is really a separate issue (I've been calling that "schema aware APIs", and that will be a separate layer on top of this).

I have also made everything explicitly read-only, since modifying these objects directly is not currently supported as a way to edit.

```suggestion

/**
 * A symbol to for the type of a node in contexts where string keys are already in use for fields.
 */
export const type: unique symbol = Symbol("editable-tree type");

 /**
 * A symbol to for the value of a node in contexts where string keys are already in use for fields.
 */
export const value: unique symbol = Symbol("editable-tree value");

/**
 * A tree which can be traversed and edited.
 *
 * When iterating, only visits non tempt fields.
 * To discover empty fields, inspect the schema using {@link "type"}.
 *
 * TODO: support editing.
 */
export interface EditableTree extends Iterable<EditableField> {
    /**
     * The type of this node.
     * If this node is is well-formed, it must follow this schema.
     */
    readonly [type]: TreeSchema;

    /**
     * Value stored on this node.
     */
    readonly [value]: Value;

    /**
     * Fields of this node, indexed by their field keys (as strings).
     *
     * This API exposes content in a way depending on the {@link Multiplicity} of the {@link FieldKind}.
     * Sequences (including empty ones) are always exposed as arrays,
     * and everything else is either a single EditableTree or undefined depending on if it's empty.
     *
     * TODO:
     * This approach to field lookup can result in collisions between global and local keys,
     * particularly with "extra" fields.
     * A mechanism for disambiguating this should be added,
     * likely involving an alternative mechanism for looking up global fields via symbols.
     */
    readonly [key: string]: readonly EditableTree[] | EditableTree | undefined;
}

/**
 * A field of an {@link EditableTree}.
 */
export type EditableField = readonly [FieldSchema, readonly EditableTree[]];

```

CraigMacomber replied:
this proposed API leaves all the primitive values boxed in their nodes, forcing use of [value] to access them.

If you want to automatically unwrap specific node types directly to their values, you could replace the use of EditiableTree for child access with:

```typescript
/**
 * EditableTree,
 * but with {@link jsonNumber}, {@link jsonBoolean} and {@link jsonString} unwrapped into just their values.
 *
 * TODO:
 * This assumes the schema being used for these trees (which could be a stored schema)
 * matches those specific Json view schema.
 * What happens if they mismatch should be defined.
 *
 * TODO:
 * EditableTree should provide easy access to children in a way thats guaranteed
 * not to do this unwrapping for cases which need to refer to the actual nodes.
 * This may include cases like creating anchors and/or editing.
 */
export type EditableTreeOrPrimitive = EditableTree | number | string | boolean;
```

sharptrip replied:
Thanks for your feedback. We are a bit confused now as the API proposed by you is a totally different thing.

Our initial goal was to have an API which mimics a JS object behaviour. If, while traversing the tree, we meet a node having children, we proxify it and dig deeper. If it is a "leaf", it should work just like a generic JS object:
```ts
assert.equal(proxy.name, "Adam");
```
If one would like to get all children (or fields) of the current node, `Object.keys(proxy)` is a way to go.
I have added another test `"traverse a complete tree"` which more or less mocks a behaviour we'd like to get for Inspector.
To be more specific, this type definition allows to have such syntax in tests like `proxy.address.street` etc. without having typecasts all the time one needs to go down to a deeper child:
```
export type EditableTreeNode<T> = {
	[P in keyof T]?: T[P] extends number | string | boolean ? T[P] : EditableTree<T[P]>;
};
```
whereas the second part (`IEditableTreeSignature`) safeguards our proxy against not allowed fields and key collisions.

We could have the proposed API implemented as well, but that wouldn't fit our needs.

sharptrip replied:
> I have also made everything explicitly read-only, since modifying these objects directly is not currently supported as a way to edit.

That's safeguarded with an assertions and tests, also we'd like to start implementing edit like yesterday =)

Overall I have an impression that it is forgotten that this PR is a WiP (in general) and 1-2 other PRs (like UI support, schematised API and edit support) will immediately follow, meanwhile improving this by using it in a "real" implementation of other features. We hoped that this would be an iterative process.

CraigMacomber commented on packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts:
```diff
@@ -0,0 +1,234 @@
+			return brand("");
+		}
+		return this.cursor.type;
+	}
```
With the API change I proposed above, this would only need to handle the case without a key.

If we do end up keeping this, the approach you use to get the type of the field here is not good: you are just returning the type of the first node, but this does not work for polymorphic fields or empty fields. It's better to get use the TreeSchema and lookup the proper field schema instead. My proposed API change exposes the actual TreeSchema under [type] so users can easily grab the schema for fields off of that if desired, and won't need a function here to do that.

CraigMacomber commented on packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts:
```diff
@@ -0,0 +1,234 @@
+	deleteProperty: (target: ProxyTarget, key: string): boolean => {
+		throw new Error("Not implemented.");
+	},
+	has: (target: ProxyTarget, key: string | symbol): boolean => {
```
I just learned that apparently you can't just user proxy to implement any API you want regardless of the target type because of these invariants for "has": https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy/Proxy/has#invariants

I guess we have to follow those which makes what properties exist on the target matter.
Maybe we should just return true for all strings to avoid violating this. Either that or we have to make the target only have symbols (Which I think you did before, but I wrongly though was pointless)

sharptrip replied:
I am not sure that it matters since
a) we don't define non-configurable properties neither define a target to be not extensible
b) as we expose the proxy via a public API type, such exception should never occur as it is hidden behind the API

The weak point I see currently is that we have to always keep the target's type and the API in sync semantically, what's always not as good as a physically implemented contract, imho, but I think that might be fine for a while.

CraigMacomber commented on packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts:
```diff
@@ -0,0 +1,234 @@
+ * and for each and every new method its own symbol should be created correspondingly.
+ * This prevents accidental mixin with user-defined string keys.
+ */
+export interface IEditableTreeSignature<T> {
```
General naming note: it turns out this repo is inconsistent with its use of the "I" prefix on interfaces, and no one really sees it as offering value, so Tree might stop using that pattern. For now, you can do whichever you want and once a decision is reached, we can go update the names accordingly. Personally, I dislike the "I" prefix, and I think we might end up not using it, which is why I didn't use it in some of my suggested code.

sharptrip committed:
Merge branch 'main' into forest-proxy

sharptrip committed:
return schema for type, add tests, fix undefined values

sharptrip committed:
add symbol index

sharptrip committed:
cleanup

sharptrip committed:
Merge branch 'main' into forest-proxy

sharptrip committed:
support empty forest

CraigMacomber committed:
use schema in EditableTree

CraigMacomber committed:
Cleanup and fixes

CraigMacomber committed:
fix tests

sharptrip committed:
Merge pull request #3 from CraigMacomber/reproxy

Tree: Use Schema in EditableTree

sharptrip committed:
cleanup array tests, export isPrimitiveValue

sharptrip committed:
restrict undefined for primitives

sharptrip committed:
Merge branch 'main' into forest-proxy

sharptrip committed:
clear instead of free for edit

CraigMacomber commented:
I think something went wrong with the latest merge from main: it introduced a large number of unrelated changes to this PR, and some merge conflicts with main.

CraigMacomber commented on packages/dds/tree/src/feature-libraries/editable-tree/utilities.ts:
```diff
@@ -12,11 +12,11 @@ import { fieldKinds } from "../defaultFieldKinds";
 
 /**
- * @returns true iff `schema` trees should default to being viewed as just their value when possible.
+ * @returns true if `schema` trees should default to being viewed as just their value when possible.
```
This was not a typo. I used [iff](https://en.wikipedia.org/wiki/If_and_only_if) on purpose so that it is unambiguous that false means not a primitive.

sharptrip committed:
[restored] types for primitives, cleanup

sharptrip commented on packages/dds/tree/src/feature-libraries/editable-tree/utilities.ts:
```diff
@@ -0,0 +1,92 @@
+import { FieldKind } from "../modular-schema";
+
+/**
+ * @returns true if `schema` trees should default to being viewed as just their value when possible.
```
```suggestion
 * @returns true iff `schema` trees should default to being viewed as just their value when possible.
```

sharptrip committed:
Update packages/dds/tree/src/feature-libraries/editable-tree/utilities.ts

CraigMacomber commented on packages/dds/tree/src/test/feature-libraries/editableTree.spec.ts:
```diff
@@ -0,0 +1,496 @@
+    }
+}
+
+describe.only("editable-tree", () => {
```
Gott make sure we don't check this in. Would skip all other tests in the package! (Looks like I left that in in my changes! Not good.)

```suggestion
describe("editable-tree", () => {
```

CraigMacomber commented on packages/dds/tree/src/test/feature-libraries/editableTree.spec.ts:
```diff
@@ -0,0 +1,496 @@
+
+function expectTreeEquals(inputField: UnwrappedEditableField, expected: JsonableTree): void {
+    assert(inputField !== undefined);
+    assert(!Array.isArray(inputField));
```
This function takes in an `UnwrappedEditableField`. An `UnwrappedEditableField` explicitly allows `readonly UnwrappedEditableTree[]`. Also, since currently some trees (those with primary sequence fields) get unwrapped into a `readonly UnwrappedEditableTree[]`, it makes sense to allows arrays here.

We have a test which covers this case "array nodes get unwrapped". Its implemented using assert.deepStrictEqual instead of this comparison function since it ended up being slightly simpler / more clear that way, but it should be possible to use expectTreeEquals in its implementation. For example the empty case in that test should be able to be changed to:

```
        // Empty
        {
            const data = { type: phonesSchema.name };
            const forest = setupForest(schemaData, [data]);
            const [context, field] = getEditableTree(forest);
            assert.deepStrictEqual(field, []);
            expectTreeEquals(field, data);
            context.free();
        }
```

and still pass, but now it would fail.

sharptrip replied:
Got your point, thanks. Fixed.

CraigMacomber commented on packages/dds/tree/src/feature-libraries/editable-tree/README.md:
```diff
@@ -0,0 +1,12 @@
+# Editable Tree
```
I recently normalized all these sub-remes to just match the directory/package name with their title exactly.
```suggestion
# editable-tree
```

sharptrip committed:
Update packages/dds/tree/src/feature-libraries/editable-tree/README.md

> I recently normalized all these sub-remes to just match the directory/package name with their title exactly.

Co-authored-by: Craig Macomber (Microsoft) <42876482+CraigMacomber@users.noreply.github.com>

sharptrip committed:
Update packages/dds/tree/src/test/feature-libraries/editableTree.spec.ts

Co-authored-by: Craig Macomber (Microsoft) <42876482+CraigMacomber@users.noreply.github.com>

sharptrip committed:
Merge branch 'main' into forest-proxy

sharptrip committed:
expect arrays upfront

sharptrip committed:
assert undefined values, tests with getTypeName, more cleanup

CraigMacomber commented on packages/dds/tree/src/test/feature-libraries/editableTree.spec.ts:
```diff
@@ -0,0 +1,511 @@
+/*!
+* Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+* Licensed under the MIT License.
+*/
```
Formatting got messed up here causing the repo policy check to fail:

```suggestion
/*!
 * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
 * Licensed under the MIT License.
 */

```

CraigMacomber commented:
Generally, looks good to me now, but since I wrote a significant amount of the code here, I'm going to ask some of my teammates to review it as well.

yann-achard-MS commented on packages/dds/tree/src/test/feature-libraries/editableTree.spec.ts:
```diff
@@ -0,0 +1,511 @@
+        const proxy = buildTestPerson();
+        assert.ok(proxy);
+        assert.equal(Object.keys(proxy).length, 5);
+        assert.strictEqual(proxy[typeSymbol], personSchema);
+        assert.strictEqual(proxy.address[typeSymbol], addressSchema);
+        assert.strictEqual((proxy.address.phones[2] as ComplexPhoneType)[typeSymbol], complexPhoneSchema);
+        assert.strictEqual(proxy[getTypeNameSymbol]("name"), stringSchema.name);
+        assert.strictEqual(proxy.address[getTypeNameSymbol]("phones"), phonesSchema.name);
```
The `import { strict as assert } from "assert";` at the top means `assert.strictEqual` is equivalent to `assert.equal`.
```suggestion
        assert.strictEqual(proxy[typeSymbol], personSchema);
        assert.equal(proxy.address[typeSymbol], addressSchema);
        assert.equal((proxy.address.phones[2] as ComplexPhoneType)[typeSymbol], complexPhoneSchema);
        assert.equal(proxy[getTypeNameSymbol]("name"), stringSchema.name);
        assert.equal(proxy.address[getTypeNameSymbol]("phones"), phonesSchema.name);
```

CraigMacomber replied:
oops, thats my mistake. Would be nice if those methods were marked as deprecated as there is no reason to call them directly.

yann-achard-MS commented on packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts:
```diff
@@ -0,0 +1,434 @@
+    getFieldKind, getFieldSchema, getPrimaryField, isPrimitive, isPrimitiveValue, PrimitiveValue,
+} from "./utilities";
+
+// Symbol for extracting target from editable-tree proxies.
+// Useful for debugging and testing, but not part of the public API.
```
```suggestion
// Symbol for extracting target from editable-tree proxies.
// Useful for debugging and testing, but not part of the public API.
```
```suggestion
// Symbol for extracting target from editable-tree proxies.
// Useful for debugging and testing, but not part of the public API.
```
```suggestion
/**
 * Symbol for extracting target from editable-tree proxies.
 * Useful for debugging and testing, but not part of the public API.
 */
```

yann-achard-MS commented on packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts:
```diff
@@ -0,0 +1,434 @@
+    getFieldKind, getFieldSchema, getPrimaryField, isPrimitive, isPrimitiveValue, PrimitiveValue,
+} from "./utilities";
+
+// Symbol for extracting target from editable-tree proxies.
+// Useful for debugging and testing, but not part of the public API.
```
```suggestion
/**
 * Symbol for extracting target from editable-tree proxies.
 * Useful for debugging and testing, but not part of the public API.
 */
```

yann-achard-MS replied:
I'm a little confused by the name: if this is used to access the target of the proxy then shouldn't it named `target` or `proxyTarget`? This is not used to access the proxy, right?

yann-achard-MS commented on packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts:
```diff
@@ -0,0 +1,434 @@
+export const proxySymbol: unique symbol = Symbol("editable-tree proxy");
+
+/**
+ * A symbol to for the type of a node in contexts where string keys are already in use for fields.
```
```suggestion
 * A symbol to get the type of a node in contexts where string keys are already in use for fields.
```

yann-achard-MS replied:
Same for other symbols below.

yann-achard-MS commented on packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts:
```diff
@@ -0,0 +1,434 @@
+/**
+ * A tree which can be traversed and edited.
+ *
+ * When iterating, only visits non tempt fields.
```
```suggestion
 * When iterating, only visits non-empty fields.
```

yann-achard-MS commented on packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts:
```diff
@@ -0,0 +1,434 @@
+
+/**
+ * EditableTree,
+ * but with for a type that `isPrimitive` unwrapped into just the value if that value is a {@link PrimitiveValue}.
```
Grammar

CraigMacomber replied:
```suggestion
 * but with any type that `isPrimitive` unwrapped into the value if that value is a {@link PrimitiveValue}.
```

yann-achard-MS commented on packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts:
```diff
@@ -0,0 +1,434 @@
+/**
+ * Unwraps fields with non-sequence multiplicities.
+ */
+export type UnwrappedEditableField = UnwrappedEditableTree | undefined | readonly UnwrappedEditableTree[];
```
Is the `readonly UnwrappedEditableTree[]` case for array nodes? If so, it would be good to indicate that in comment as the presence of `readonly UnwrappedEditableTree[]` is surprising here.

CraigMacomber replied:
That case is for sequence fields. I have added a suggestion above for clarifying this comment.

yann-achard-MS commented on packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts:
```diff
@@ -0,0 +1,434 @@
+     */
+    public getTypeName(key: string): TreeSchemaIdentifier {
+        const childTypes = mapCursorField(this.cursor, brand(key), (c) => c.type);
+        assert(childTypes.length <= 1, "invalid non sequence");
```
```suggestion
        assert(childTypes.length <= 1, "invalid on sequence");
```

yann-achard-MS commented on packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts:
```diff
@@ -0,0 +1,434 @@
+
+	public get cursor(): ITreeSubscriptionCursor {
+		if (this.lazyCursor.state === ITreeSubscriptionCursorState.Cleared) {
+            assert(this.anchor !== undefined, "EditableTree should have an anchor if it does not have a cursor");
+			const result = this.context.forest.tryMoveCursorTo(this.anchor, this.lazyCursor);
+			assert(result === TreeNavigationResult.Ok,
+                "It is invalid to access an EditableTree node which no longer exists");
+            this.context.withCursors.add(this);
+            this.context.forest.anchors.forget(this.anchor);
+            this.context.withAnchors.delete(this);
+            this.anchor = undefined;
```
Is there a way we can add test coverage for this?

CraigMacomber replied:
We should at minimum include a TODO (here or in the test suite) to cover ProxyContext's methods.

sharptrip replied:
Yes, it will definitely be covered e.g. in https://github.com/sharptrip/FluidFramework/pull/4. I'll add a TODO for now.

CraigMacomber commented on packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts:
```diff
@@ -0,0 +1,434 @@
+export type EditableField = readonly [FieldSchema, readonly EditableTree[]];
+
+/**
+ * Unwraps fields with non-sequence multiplicities.
```
```suggestion
 * Unwrapped field.
 * Non-sequence multiplicities are unwrapped to the child tree or `undefined` if there is none.
 * Sequence multiplicities are handled with arrays.
 * See {@link UnwrappedEditableTree} for how the children themselves are unwrapped.
```

CraigMacomber commented on packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts:
```diff
@@ -0,0 +1,434 @@
+        this.context.withCursors.delete(this);
+    }
+
+	public get cursor(): ITreeSubscriptionCursor {
```
mixed tabs and spaces. We use 4 space indent in this project currently. Please convert the indentation to spaces.

sharptrip replied:
probably, vs code issue. fixed.

CraigMacomber commented on packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts:
```diff
@@ -0,0 +1,434 @@
+			assert(result === TreeNavigationResult.Ok,
+                "It is invalid to access an EditableTree node which no longer exists");
+            this.context.withCursors.add(this);
+            this.context.forest.anchors.forget(this.anchor);
+            this.context.withAnchors.delete(this);
+            this.anchor = undefined;
```
```suggestion
```
There is no reason to delete the anchor here. While the anchor is redundant, we might as well keep it.
If this proxy is cleaned up before the next edit, we can leave forgetting it for then. If its not cleaned up by then, we will have to create the anchor again, and keeping here would avoid that.

CraigMacomber replied:
That cost is incurred when you edit.  If you throw away the anchor, you will have to recreate it before you edit, and thus incur the cost of having an anchor, and the cost of recreating it.

Therefor throwing the anchor away here increases the CPU cost. The only potential win here is using slightly less memory between now and the next edit (when you have to reallocate the anchor) or this node is disposed (and you would throw away the anchor anyway).

Also I don't see how throwing away the anchor would help implement editing code. The anchor for a given proxy is never going to mutate, and we simply compute it lazily for cost reasons. Editing code will want to use the anchors: making them go away between edits seems harmful.

CraigMacomber commented on packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts:
```diff
@@ -0,0 +1,434 @@
+        return getFieldKind(getFieldSchema(this.type, key));
+    }
+
+    get keys(): string[] {
```
Because this is a getter, having it return a mutable array is confusing (a caller might expect modifying it to update the object). I suggest either making this a readonly array, or not making this a getter and using a normal method instead.

CraigMacomber commented on packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts:
```diff
@@ -0,0 +1,434 @@
+            targets.push(new ProxyTarget(context, cursor));
+        } while (cursor.seek(1) === TreeNavigationResult.Ok);
+    }
+    cursor.clear();
```
this is unneeded. If we wanted to reuse this cursor later, we could clear it, but since its just getting freed below this, there is no benifit, and just added overhead.
```suggestion
```

sharptrip replied:
But then it will stay in a forest forever, isn't it? Meaning also that you can't edit the forest neither clear this cursor from outside the forest API, looks like a dead end.

CraigMacomber replied:
Clearing allows you to edit despite the cursor not being freed (edits the cursor to remove its position). Free deletes the cursor from the forest (making its position not matter), which also allows you to edit. Since the line after this is free, clearing is unneeded.

sharptrip replied:
I have implemented some tests for edits in https://github.com/sharptrip/FluidFramework/pull/4. If this line is removed, I immediately get `Error: 0x374` from the ObjectForest.

I think it does not work as you said, at least, I don't see where `free()` deletes the cursor:

https://github.com/microsoft/FluidFramework/blob/91153dd3359cee47966e7927db8a2df8d267a980/packages/dds/tree/src/feature-libraries/object-forest/objectForest.ts#L242-L245

CraigMacomber replied:
That is a bug in object forest. I'll fix it and clean up this extra clear call when I do. Good find!

CraigMacomber replied:
I pushed a fix for this.

sharptrip replied:
> I pushed a fix for this.

Thanks! That's basically my problem here: it's still hard to say, if there is a bug or a misuse from my side.

CraigMacomber commented on packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts:
```diff
@@ -0,0 +1,434 @@
+ */
+export const value: unique symbol = Symbol("editable-tree value");
+
+/**
+ * A symbol to for the type name of a child node in contexts where string keys are already in use for fields.
+ */
+ export const getTypeName: unique symbol = Symbol("editable-tree typeName");
```
I'd prefer to avoid having this.

I want to encourage everyone at this abstraction level to use the schema object instance instead of the type name. This makes handling stored vs view schema easier (you never can accidently use the type name to look up the schema and get the wrong one). Its also more forwar
```suggestion
/**
 * A symbol to for the type name of a child node in contexts where string keys are already in use for fields.
 *
 * TODO:
 * Add name to TreeSchema retuned via `type` and remove this.
 * We want to encourage using the TreeSchema instances and not the string names to improve robustness when multiple schema have the same name (for example a stored schema and a view schema which mismatch, or when if later add support for contextual schema).
 * Also using the schema instance directly should be faster for a lot of use cases (reference comparison should be faster than string comparison, and access to data from schema will avoid a map lookup)
 */
 export const getTypeName: unique symbol = Symbol("editable-tree typeName");
```

CraigMacomber replied:
Thats probably going to need some line wrapping to not fail the linter.

sharptrip replied:
As for the Inspector usecase, we actually don't need anything other than a type name for UI. A field schema might be needed later, but we need build-in options to narrow down this complexity for simple usecases. I'll convert into a function and add some options.

sharptrip replied:
Another issue is to get types for primitive fields. So, anyway it should be a function with at least a key to get types of the children.

CraigMacomber replied:
I intended that usecase (getting types of children) to be covered by a couple different approaches:

* If you want the schema for the field (what types are allowed there), you can get the schema for the node and look up the type of the field. This already works (though due to annoyances with local vs global field key aliasing, it's not a great API yet).
* If you want the type of the child node, we will provide an API that lets you get the child node without unwrapping (ex, maybe via a "getWrappedChild" symbol or something). There is a TODO to add such an API on EditableTree (copied below).
* We could also provide an API that lets you access the field without unwrapping. For now this would just give you an array, but once we have field proxies, there could be an API on it to get the field schema (as well as an API to get children with or without unwrapping). I think this also falls under the below TODO.

```
TODO: provide non-schema impacted APIs for getting fields and nodes without unwrapping
 * (useful for generic code, and when references to these actual fields and nodes are required,
 * for example creating anchors and editing).
```

CraigMacomber commented on packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts:
```diff
@@ -0,0 +1,434 @@
+function inProxyOrUnwrap(target: ProxyTarget): UnwrappedEditableTree {
+    if (isPrimitive(target.type)) {
+        const nodeValue = target.cursor.value;
+        // see `isPrimitive` doc for `undefined` case
+        assert(nodeValue !== undefined, "Undefined values not allowed");
+        if (isPrimitiveValue(nodeValue)) {
```
This assert is not valid. A node which has ValueSchema.Serializable can have an undefined value here, and we need to support that.

How that should be handled is isPrimitiveValue should return false when given undefined (or should have its type signature adjusted to now allow passing in undefined).

Looks like the current implementation of isPrimitiveValue has a bug where it returns true when given undefined: fix that, and remove this assert and it should work properly.

```suggestion
        if (isPrimitiveValue(nodeValue)) {
```

sharptrip replied:
To be honest, I thought that `Serializable` was omitted on purpose for this primitive node
https://github.com/microsoft/FluidFramework/pull/11589/files?diff=unified&w=1#diff-ba26d1ea70456ae7a727078e93a12eb306cd25117730f41aa9aeca3093f80c32R71-R74

The problem is imho if one will accidentally provide `undefined` instead of a primitive value, and this will be somewhere deeply in the tree, it will be very troublesome to figure out what happens, as without any assertion (not necessarily this one) this code will silently return a proxy.

I've reworked this to assert if `undefined` node is `Serializable` (and fixed `isPrimitiveValue` as suggested below), but I don't like to allow undefined numbers or strings as this potentially could lead to an undefined behaviour.

CraigMacomber replied:
There are two definitions of "primitive" here. One at the schema level which just has a placeholder implementation (isPrimitive) which will eventually be replaced by metadata on the view schema (I intentionally made this overly general, so EditiableTree needs to handle cases it can't unwrap so that future changes to how this works won't break editiableTree), and a more strict definition at the node level that is logically part of the EditiableTree implementation which explicitly excludes undefined and objects from being treated as primitives.

Note that if you use ValueSchema's Number, Boolean or String, undefined is not allowed (would be out of schema), and thus you will unconditionally pass both checks and end up fully unwrapped.

This proxy can assume the data is in schema, and can (due does not have to) assert on out of schema cases. This assert could however be hit by data that is in schema and uses Serializable which is the problem.

The edge case here is that if the schema explicitly says Serializable, EditableTree need to handle all the allowed values for that (actual primitives, undefined, and objects). Since the tests currently arn't using Serializable, I decided it was fine for these cases to unwrap the actual primitive values, but keep the other cases (undefined and objects) in proxies. We can change this policy as long as we actually still handle all in schema cases without asserting.

I don't see any undefined behavior here: I think the behavior for this is defined in EditiableTreeOrPrimative's documentation.

sharptrip replied:
Em, I don't think you got my last point. With

> I've reworked this...

I mean, I have changed the code, so that the assert you mentioned initially is not there anymore, and the rest is exactly about to let `undefined` go throw if `Serializable` is explicitly defined in a schema, not letting primitive nodes with primitive values to be unwrapped accidentally just because someone put `undefined` into their value.

CraigMacomber replied:
I was not objecting to your change, just clarifying why I set it up how I did. Sorry for the confusion.

CraigMacomber commented on packages/dds/tree/src/feature-libraries/editable-tree/utilities.ts:
```diff
@@ -0,0 +1,92 @@
+export type PrimitiveValue = string | boolean | number;
+
+export function isPrimitiveValue(nodeValue: Value): nodeValue is PrimitiveValue {
+    return typeof nodeValue !== "object";
```
```suggestion
    return nodeValue !== undefined && typeof nodeValue !== "object";
```

CraigMacomber replied:
Should probably have tests for this function. Adding a utilities.spec.ts for tests for things in this file would be great.

sharptrip replied:
done

sharptrip committed:
format indents

sharptrip committed:
Update packages/dds/tree/src/test/feature-libraries/editableTree.spec.ts

formatting

Co-authored-by: Craig Macomber (Microsoft) <42876482+CraigMacomber@users.noreply.github.com>

sharptrip committed:
Replace strictEqual with just equal

sharptrip commented on packages/dds/tree/src/test/feature-libraries/editableTree.spec.ts:
```diff
@@ -0,0 +1,511 @@
+    extraLocalFields: emptyField,
+});
+
+// This schema is really unnecessary: it could just use a sequence field instead.
```
I guess this comment is about "explicit arrays" vs "implicit sequences" from https://github.com/microsoft/FluidFramework/tree/main/packages/dds/tree/docs/data-model#explicit-arrays

sharptrip committed:
Fix undefined for primitives, merge type and getTypeName, cleanup

sharptrip committed:
more cleanup

sharptrip committed:
Comments for EditableTreeContext

sharptrip commented on packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts:
```diff
@@ -0,0 +1,434 @@
+ */
+export type UnwrappedEditableField = UnwrappedEditableTree | undefined | readonly UnwrappedEditableTree[];
+
+export interface EditableTreeContext {
```
done

sharptrip commented on packages/dds/tree/src/test/feature-libraries/editableTree.spec.ts:
```diff
@@ -0,0 +1,511 @@
+const optionalChildSchema = namedTreeSchema({
+    name: brand("Test:OptionalChild-1.0.0"),
+    localFields: {
+        child: fieldSchema(FieldKinds.optional),
```
@CraigMacomber could you please comment here?

jenn-le replied:
Ahh I see. Thanks, Craig!

sharptrip committed:
Merge branch 'main' into forest-proxy

sharptrip committed:
re-build api report, more cleanup

sharptrip committed:
test utilities

CraigMacomber commented on packages/dds/tree/src/test/feature-libraries/editableTree.spec.ts:
```diff
@@ -0,0 +1,511 @@
+        friends: fieldSchema(FieldKinds.value, [mapStringSchema.name]),
+        address: fieldSchema(FieldKinds.value, [addressSchema.name]),
+    },
+    extraLocalFields: emptyField,
```
This is due to an implementation constraint. The default value you want to use for extraLocalFields depends on the defaultFieldKinds, which are out of scope for where the fieldSchema function is implemented. Once field kinds are more baked into the system, we can write a more proper schema language that takes a hard dependency on the specific field kinds and fixes these kinds of issues.

sharptrip committed:
Update packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts

Co-authored-by: Craig Macomber (Microsoft) <42876482+CraigMacomber@users.noreply.github.com>

sharptrip commented on packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts:
```diff
@@ -0,0 +1,441 @@
+     */
+    readonly [valueSymbol]: Value;
+
+    readonly [proxyTargetSymbol]: object;
```
```suggestion
    /**
     * Stores {@link ProxyTarget}, which implements reading and writing for this node.
     */
    readonly [proxyTargetSymbol]: object;
```

sharptrip committed:
Update packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts

CraigMacomber commented on packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts:
```diff
@@ -0,0 +1,444 @@
+     */
+    readonly [valueSymbol]: Value;
+
+    /**
+     * Stores {@link ProxyTarget}, which implements reading and writing for this node.
+     */
+    readonly [proxyTargetSymbol]: object;
```
API Extractor is objecting to a link to a non-exported type (that would be a broken link in the docs).
```suggestion
    /**
     * Stores the target for the proxy which implements reading and writing for this node.
     * The details of this object are implementation details,
     * but the presence of this symbol can be used to separate EditableTrees from other types.
     */
    readonly [proxyTargetSymbol]: object;
```

CraigMacomber committed:
Update packages/dds/tree/src/feature-libraries/editable-tree/editableTree.ts

CraigMacomber committed:
update api report

CraigMacomber committed:
fix object forest bug with cursor freeing
